ulib/LowStar.Vector.fst

```diff
@@ -0,0 +1,687 @@
+
+val back:
+  #a:Type -> vec:vector a{size_of vec > 0ul} ->
+  HST.ST a
+    (requires (fun h0 -> live h0 vec))
+    (ensures (fun h0 v h1 -> 
+      h0 == h1 /\ S.index (as_seq h1 vec) (U32.v (size_of vec) - 1) == v))
+let back #a vec =
+  B.index (Vec?.vs vec) (size_of vec - 1ul)
+
+/// Operations
+
+val clear:
+  #a:Type -> vec:vector a ->
+  Tot (cvec:vector a{size_of cvec = 0ul})
+let clear #a vec =
+  Vec 0ul (Vec?.cap vec) (Vec?.vs vec)
+
+val clear_as_seq_empty:
+  #a:Type -> h:HS.mem -> vec:vector a ->
+  Lemma (S.equal (as_seq h (clear vec)) S.empty)
+	[SMTPat (as_seq h (clear vec))]
+let clear_as_seq_empty #a h vec = ()
+
+private val slice_append:
+  #a:Type -> s:S.seq a ->
+  i:nat -> j:nat{i <= j} -> k:nat{j <= k && k <= S.length s} ->
+  Lemma (S.equal (S.slice s i k)
+		 (S.append (S.slice s i j) (S.slice s j k)))
+private let slice_append #a s i j k = ()
+
+val assign:
+  #a:Type -> vec:vector a ->
+  i:uint32_t -> v:a ->
+  HST.ST unit
+    (requires (fun h0 -> live h0 vec /\ i < size_of vec))
+    (ensures (fun h0 _ h1 ->
+      hmap_dom_eq h0 h1 /\
+      modifies (loc_vector_within #a vec i (i + 1ul)) h0 h1 /\
+      get h1 vec i == v /\
+      S.equal (as_seq h1 vec) (S.upd (as_seq h0 vec) (U32.v i) v)))
+#reset-options "--z3rlimit 10"
+let assign #a vec i v =
+  let hh0 = HST.get () in
+  // NOTE: `B.upd (Vec?.vs vec) i v` makes more sense, 
+  //       but the `modifies` postcondition is coarse-grained.
+  B.upd (B.sub (Vec?.vs vec) i 1ul) 0ul v;
+  let hh1 = HST.get () in
+  loc_vector_within_disjoint vec 0ul i i (i + 1ul);
+  modifies_as_seq_within 
+    vec 0ul i (loc_vector_within #a vec i (i + 1ul)) hh0 hh1;
+  loc_vector_within_disjoint vec i (i + 1ul) (i + 1ul) (size_of vec);
+  modifies_as_seq_within 
+    vec (i + 1ul) (size_of vec) (loc_vector_within #a vec i (i + 1ul)) hh0 hh1;
+  slice_append (as_seq hh1 vec) 0 (U32.v i) (U32.v i + 1);
+  slice_append (as_seq hh1 vec) 0 (U32.v i + 1) (U32.v (size_of vec));
+  slice_append (S.upd (as_seq hh0 vec) (U32.v i) v) 0 (U32.v i) (U32.v i + 1);
+  slice_append (S.upd (as_seq hh0 vec) (U32.v i) v) 0 (U32.v i + 1) (U32.v (size_of vec))
+
+private val resize_ratio: uint32_t
+private let resize_ratio = 2ul
+
+private val new_capacity: cap:uint32_t{cap > 0ul} -> Tot uint32_t
+private let new_capacity cap =
+  if cap >= max_uint32 / resize_ratio then max_uint32
+  else cap * resize_ratio
+
+val insert: 
+  #a:Type -> vec:vector a -> v:a -> 
+  HST.ST (vector a)
+    (requires (fun h0 -> 
+      live h0 vec /\ freeable vec /\ not (is_full vec) /\
+      HST.is_eternal_region (frameOf vec)))
+    (ensures (fun h0 nvec h1 ->
+      frameOf vec = frameOf nvec /\
+      hmap_dom_eq h0 h1 /\
+      live h1 nvec /\ freeable nvec /\
+      modifies (loc_union (loc_addr_of_vector vec) 
+      			  (loc_vector nvec)) h0 h1 /\
+      size_of nvec = size_of vec + 1ul /\
+      get h1 nvec (size_of vec) == v /\
+      S.equal (as_seq h1 nvec) (S.snoc (as_seq h0 vec) v)))
```

Comment:
@joonwonc Sorry for the slow reviews on this ...

I see now that `insert` resizes the vector, but it does by reallocating and passing back a resized vector, rather than mutating the original.

This is fine, but it is not the design goal described to me by @ad-l @protz and @fournet 

Is this expected by all?

Reply:
I got your point, two things to mention here:

1. If you meant to maintain the original vector **struct** but to reallocate a **buffer**, this is possible by passing a vector struct **pointer** to the function and changing field values appropriately. As I remember, at the time I built this library we didn't have a way to use a field access with a struct pointer (`->`), so it currently returns a new vector struct. But as far as I know we can use `->` now so I could change the interface.

2. If you meant to maintain the original **buffer** pointer, we need a way to change the length of an allocated buffer. As far as I know we don't have such mechanism in C. So we just need to allocate a new buffer with a new capacity and to copy the content. This mechanism is also used by C++ std::vector, and for example, because of that, whenever we add en element to a vector, all iterators for the vector are invalidated (https://en.cppreference.com/w/cpp/container/vector/push_back).

Reply:
Passing a pointer to a struct then relying on KreMLin's pretty-printing to compile:

```
p.(0ul) <- { p.(0ul) with underlying_buffer with the_newly_allocated_buffer }
```

as:
```
p->underlying_buffer = the_newly_allocated_buffer
```
is always feasible on top of the current API, I believe (possibly with suitable inlining). However, this would be quite a burden to clients of this API, having to manage the outer pointer.

Since the struct is (I believe) three words in C, I would assume that making this function a struct-returning one, rather than an in-place mutating one, won't have too much of a performance impact.

I remember having similar discussions with @tahina-pro and also advocating for the struct-returning style, in order to limit the burden on clients.

So, in short, the current style is fine for me, and we can always define the more C-idiomatic style on top of it, should we want to.

Reply:
I don't see why the client should have to explicitly manipulate a pointer to a struct with reliance on KreMLin for better code etc.

I was expecting the library to internally represent a vector as a pointer to a buffer, its current size and max capacity ... with buffer like operations (index, update) etc. lifted to this representation, and insert possibly resizing the buffer in case we're out of capacity. This should be completely transparent to clients.